WalletWasabi/Wallets/FileSystemBlockRepository.cs

```diff
@@ -15,17 +16,124 @@ namespace WalletWasabi.Wallets
 	/// </summary>
 	public class FileSystemBlockRepository : IRepository<uint256, Block>
 	{
-		public FileSystemBlockRepository(string blocksFolderPath, Network network)
+		public FileSystemBlockRepository(string blocksFolderPath, Network network, long targetBlocksFolderSizeMb = 300)
 		{
 			BlocksFolderPath = blocksFolderPath;
 			Network = network;
 			CreateFolders();
+			EnsureBackwardsCompatibility();
+			Prune(targetBlocksFolderSizeMb);
 		}
 
 		public string BlocksFolderPath { get; }
 		public Network Network { get; }
 		private AsyncLock BlockFolderLock { get; } = new AsyncLock();
 
+		private void EnsureBackwardsCompatibility()
+		{
+			try
+			{
+				// Before Wasabi 1.1.13
+				var wrongBlockFolderPath = Path.Combine(EnvironmentHelpers.GetDataDir(Path.Combine("WalletWasabi", "Client")), $"Blocks{Network}");
+				if (Directory.Exists(wrongBlockFolderPath))
+				{
+					using (BenchmarkLogger.Measure())
+					{
+						var cntSuccess = 0;
+						var cntRedundant = 0;
+						var cntFailure = 0;
+						foreach (var filePath in Directory.EnumerateFiles(wrongBlockFolderPath))
+						{
+							try
+							{
+								var fileName = Path.GetFileName(filePath);
+								var migrationPath = Path.Combine(BlocksFolderPath, fileName);
+
+								// Unintuitively File.Move overwrite: false throws an IOException if the file already exists.
+								// https://docs.microsoft.com/en-us/dotnet/api/system.io.file.move?view=netcore-3.1
+								if (!File.Exists(migrationPath))
+								{
+									File.Move(filePath, migrationPath, overwrite: false);
+									cntSuccess++;
+								}
+								else
+								{
+									cntRedundant++;
+								}
+							}
+							catch (Exception ex)
+							{
+								Logger.LogDebug(ex);
+								cntFailure++;
+							}
+						}
+
+						Directory.Delete(wrongBlockFolderPath, recursive: true);
```

Comment:
Do we have magic delete at our library?

Reply:
I thought about that, but that operates with catching random failures' exceptions and retry on delay. Here I catch the exceptions and failure of this should not delay the wallet start.